Added small note when two different closures fail typechecking.

ebfull opened the pull request:
Also fixed bug calling .note() instead of .help()

See #24036

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @brson (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

brson commented:
Thanks. Improved diagnostics are always awesome. Can you add a test case that captures this scenario?

ebfull commented:
Changed it. It should make the note/help like this now:

```
src/main.rs:13:9: 13:18 error: mismatched types:
 expected `[closure src/main.rs:12:17: 12:26]`,
    found `[closure src/main.rs:13:9: 13:18]`
(expected closure,
    found a different closure) [E0308]
src/main.rs:13     x = |c| c + 1;
                       ^~~~~~~~~
src/main.rs:13:9: 13:18 note: no two closures, even if identical, have the same type
src/main.rs:13     x = |c| c + 1;
                       ^~~~~~~~~
src/main.rs:13:9: 13:18 help: consider boxing your closure and/or using it as a trait object
src/main.rs:13     x = |c| c + 1;
                       ^~~~~~~~~
```

Added a test too. Should I rebase?

pnkfelix commented:
@bors  r+ rollup

bors commented:
:pushpin: Commit 3308c06 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 3308c06e33855f89627b417fd6b57f76d9df50a1 -->

bors commented:
:hourglass: Testing commit 3308c06 with merge 37cb1d4...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/4452), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/4459), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4424), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/4440), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4408), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/4475), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/4472), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4451), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/4140), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/4127), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/2945), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2964)
